### PR TITLE
rules/ refactoring

### PR DESCRIPTION
**Pull request by fabxc**
@brian-brazil @beorn7 

This refactors a good chunk of the rules package in preparation for the evaluation groups we eventually want. For now everything remains in a default group which's execution is still concurrent.

As in previous attempt at the retrieval package we no longer try to update running components but destroy and recreate with the required state. Tested this a fair bit. Unless there's huge contention (i.e. server basically overloaded and intervals are not nearly precise anymore anyway) there's a negligible chance of skipping an iteration. 

I will add some tests for `Group` later, but the code is good for review.

### CODE REVIEW

**Comment by brian-brazil** [comment]
:+1:

File: rules/alerting.go [hunk 1]
```diff
@@ -46,52 +46,23 @@ func (s AlertState) String() string {
...
-		panic("undefined")
 	}
+	panic(fmt.Errorf("unknown alert state: %v", s))
 }
 
 const (
-	// StateInactive is the state of an alert that is either firing nor pending.
 	StateInactive AlertState = iota
-	// StatePending is the state of an alert that has been active for less than
-	// the configured threshold duration.
```
Comment (grobie) [inline rules/alerting.go]: Why did you remove all comments? It's not clear for a new reader to know what a pending alert is, that requires knowledge we can also simply comment here.

**Comment by beorn7** [comment]
Scheduled for review... Please don't merge yet.

File: rules/alerting.go [hunk 2]
```diff
@@ -133,65 +110,146 @@ func (rule *AlertingRule) Name() string {
 	return rule.name
 }
 
+func (r *AlertingRule) sample(alert *Alert, ts model.Time, set bool) *model.Sample {
+	// Build alert labels in order they can be overwritten.
```
Comment (beorn7) [inline rules/alerting.go]: What does this comment mean? (Grammar off?)

File: rules/alerting.go [hunk 3]
```diff
@@ -133,65 +110,146 @@ func (rule *AlertingRule) Name() string {
...
 	// Check if any pending alerts should be removed or fire now. Write out alert timeseries.
-	for fp, activeAlert := range rule.activeAlerts {
+	for fp, a := range r.active {
 		if _, ok := resultFPs[fp]; !ok {
-			vector = append(vector, activeAlert.sample(timestamp, 0))
-			delete(rule.activeAlerts, fp)
+			if a.State != StateInactive {
+				vec = append(vec, r.sample(a, ts, false))
+			}
+			// If the alert was previously firing, keep it aroud for a given
```
Comment (beorn7) [inline rules/alerting.go]: aroud -> around

**Comment by fabxc** [comment]
@beorn7 was this a brief look over it or a review?

**Comment by brian-brazil** [comment]
:+1:

File: notification/notification.go [hunk 4]
```diff
@@ -159,6 +159,11 @@ func (n *Handler) nextBatch() []*model.Alert {
 
 // Run dispatches notifications continuously.
 func (n *Handler) Run() {
+	// Just warn one in the beginning to prevent nosiy logs.
```
Comment (beorn7) [inline notification/notification.go]: Typo: nosiy

Comment (beorn7) [inline notification/notification.go]: also one -> once

**Comment by beorn7** [comment]
I guess I'm happy once you have addressed my comments.

**Comment by fabxc** [comment]
:tada:
